src/vision/tenets.md

```diff
@@ -0,0 +1,20 @@
+# Design tenets for async
+
+The design tenets describe the key principles that drive our work on async. Hopefully, we are able to achieve and honor all of them all of the time. Sometimes, though, they come into conflict, and we have to pick -- in that case, we prefer the tenet earlier in the list.
+
+1. **Minimal overhead.** Rust Async I/O performance should compare favourably with any other language. In the extreme case, it should be possible to use async/await without any required allocation, although this is unlikely to be a common case in production systems.
+2. **Async is like sync, but with blocking points clearly identified.** At the highest level, writing a simple program using asynchronous I/O in Rust should be analogous to writing one that uses synchronous I/O, except that one adds `async` in front of function declarations and adds `.await` after each call. We should aim for analogous design between synchronous and asynchronous equivalents. Similarly, streams should be like asynchronous iterators. One should be able to use the same sort of combinators with streams and to iterate over them in analogous ways.
+3. **Easy to get started, but able to do anything you want.** We should make it simple to write Async I/O code and get things that work reasonably well, but it should be possible for people to obtain fine-grained control as needed.
+4. **No one true runtime.** We need to be able to hook into existing runtimes in different environments, from embedded environments to runtimes like node.js. Specialized systems need specialized runtimes. 
```

Comment:
What tradeoffs are we willing to make to achieve this runtime flexibility? The current state of runtimes results in siloed ecosystems where lots of functionality is only available within a specific runtime. Concretely: can we have a goal that says "Most users should be able to switch runtimes without rewriting their applications"?

(Related to the below point)

Reply:
I definitely believe that should be a goal. I agree it's a missing tenet.